File: shared/main/java/net/ccbluex/liquidbounce/utils/PPSCounter.java
package net.ccbluex.liquidbounce.utils;

public class PPSCounter
{
	private static final int MAX_PPS = 10000;
	private static final RollingArrayLongBuffer[] TIMESTAMP_BUFFERS = new RollingArrayLongBuffer[2];

	static
	{
		for (int i = 0, j = TIMESTAMP_BUFFERS.length; i < j; i++)
			TIMESTAMP_BUFFERS[i] = new RollingArrayLongBuffer(MAX_PPS);
	}

	/**
	 * Registers a mouse button click
	 *
	 * @param boundType
	 *               The packet type
	 */
	public static void registerPacket(final BoundType boundType)
	{
		TIMESTAMP_BUFFERS[boundType.index].add(System.currentTimeMillis());
	}

	/**
	 * Gets the count of clicks that have occurrence since the last 1000ms
	 *
	 * @param  boundType
	 *                The packet type
	 * @return        The PPT
	 */
	public static int getPacketCount(final BoundType boundType, final long time)
	{
		return TIMESTAMP_BUFFERS[boundType.index].getTimestampsSince(System.currentTimeMillis() - time);
	}

	public enum BoundType
	{
		INBOUND(0),
		OUTBOUND(1);

		final int index;

		BoundType(final int index)
		{
			this.index = index;
		}
	}
}

File: shared/main/java/net/ccbluex/liquidbounce/utils/PPSCounter.kt
package net.ccbluex.liquidbounce.utils

object PPSCounter
{
	private const val MAX_PPS = 2048
	private val TIMESTAMP_BUFFERS = arrayOfNulls<RollingArrayLongBuffer>(BoundType.values().size)

	/**
	 * Registers a mouse button click
	 *
	 * @param boundType
	 * The packet type
	 */
	@JvmStatic
	fun registerPacket(boundType: BoundType)
	{
		TIMESTAMP_BUFFERS[boundType.index]?.add(System.currentTimeMillis())
	}

	/**
	 * Gets the count of clicks that have occurrence since the last 1000ms
	 *
	 * @param  boundType
	 * The packet type
	 * @return        The PPT
	 */
	fun getPacketCount(boundType: BoundType, time: Long): Int = TIMESTAMP_BUFFERS[boundType.index]?.getTimestampsSince(System.currentTimeMillis() - time) ?: 0

	enum class BoundType(val index: Int)
	{
		INBOUND(0),
		OUTBOUND(1);
	}

	init
	{
		for (i in TIMESTAMP_BUFFERS.indices) TIMESTAMP_BUFFERS[i] = RollingArrayLongBuffer(MAX_PPS)
	}
}
